File: library/src/main/java/com/custom/photoView/photoTap/ScrollerProxy.java
package com.custom.photoView.photoTap;

import android.content.Context;
import android.os.Build;

import com.custom.photoView.photoTap.GingerScroller;
import com.custom.photoView.photoTap.IcsScroller;
import com.custom.photoView.photoTap.PreGingerScroller;

public abstract class ScrollerProxy {

    public static ScrollerProxy getScroller(Context context) {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.GINGERBREAD) {
            return new PreGingerScroller(context);
        } else if (Build.VERSION.SDK_INT < Build.VERSION_CODES.ICE_CREAM_SANDWICH) {
            return new GingerScroller(context);
        } else {
            return new IcsScroller(context);
        }
    }

    public abstract boolean computeScrollOffset();

    public abstract void fling(int startX, int startY, int velocityX, int velocityY, int minX, int maxX, int minY,
                               int maxY, int overX, int overY);

    public abstract void forceFinished(boolean finished);

    public abstract boolean isFinished();

    public abstract int getCurrX();

    public abstract int getCurrY();


}

File: library/src/main/java/com/custom/photoView/photoTapListener/ScrollerProxy.kt
package com.custom.photoView.photoTapListener

import android.content.Context
import android.os.Build

import com.custom.photoView.photoTapListener.GingerScroller
import com.custom.photoView.photoTapListener.IcsScroller
import com.custom.photoView.photoTapListener.PreGingerScroller

abstract class ScrollerProxy {

    companion object{
        fun getScroller(context: Context): ScrollerProxy {
            return if (Build.VERSION.SDK_INT < Build.VERSION_CODES.GINGERBREAD) {
                PreGingerScroller(context)
            } else if (Build.VERSION.SDK_INT < Build.VERSION_CODES.ICE_CREAM_SANDWICH) {
                GingerScroller(context)
            } else {
                IcsScroller(context)
            }
        }

    }


    abstract fun computeScrollOffset(): Boolean

    abstract fun fling(startX: Int, startY: Int, velocityX: Int, velocityY: Int, minX: Int, maxX: Int, minY: Int,
                       maxY: Int, overX: Int, overY: Int)

    abstract fun forceFinished(finished: Boolean)

    abstract fun isFinished(): Boolean

    abstract fun getCurrX(): Int

    abstract fun getCurrY(): Int


}
